File: Data-Structures-and-Algorithms-Specialization/1. Algorithmic Toolbox/Week 2/7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.cpp
#include <iostream>
#include <vector>
#include <numeric>
using namespace std;

int fibonacci_sum_last_digit(int m, int n) {
    vector<int> f;
    f.push_back(0);
    f.push_back(1);
    for (int i = 2; i <= 60; i++){
        f.push_back((f[i-1] + f[i-2]) % 10);
    }
    int rem1 = m % 60;
    int quotient1 = (m - rem1) / 60;
    int rem2 = n % 60;
    int quotient2 = (n - rem2) / 60;
    int sum1 = accumulate(f.begin() + rem1, f.end(), 0);
    int sum2 = accumulate(f.begin(), f.begin() + rem2 + 1, 0);
    int sum3 = accumulate(f.begin(), f.end(), 0) *  (quotient2 - (quotient1 + 1));
    return (sum1 + sum2 +sum3) % 10; 
}

int main() {
    int m, n;
    cin >> m >> n;
    cout << fibonacci_sum_last_digit(m, n) << '\n';
    return 0;
}

File: Data-Structures-and-Algorithms-Specialization/1. Algorithmic Toolbox/Week 2/7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.py
def calc_fib(m, n):

    f = [0, 1]
    for i in range(2, 61):
        f.insert(i, (f[i-1] + f[i-2]) % 10)
    rem1 = m % 60
    quotient1 = (m - rem1) / 60
    rem2 = n % 60
    quotient2 = (n - rem2) / 60
    sum1, sum2, sum3 = sum(f[rem1:]), sum(f[:rem2+1]), sum(f) * (quotient2 - (quotient1 + 1))
    return int((sum1 + sum2 +sum3) % 10)

if __name__ == '__main__':
    m, n = map(int, input().split())
    print(calc_fib(m, n))
